### PR TITLE
SMS + Twilio + JWT Instructions

### PR DESCRIPTION
**Pull request by peterpme**
:wave: Thanks for creating an awesome new tool! I'm LOVING Schema Extensions!

Here's my first stab at creating the flow for SMS authentication:

- Enter phone number
- Generate JWT + 5 digit code
- Text 5 digit code, save JWT to User

Any feedback would be appreciated!

### CODE REVIEW

**Comment by kbrandwijk** [comment]
I have looked at the sources. Do I understand correctly that the sms code verification itself is done on the client? If so, the client needs to be able to decrypt the token? It seems I'm missing that step somehow, or I might have overlooked/misunderstood.

**Comment by kbrandwijk** [comment]
Also, twilio offers the entire SMS verification process, so no need to work with tokens and store anything in your own database.

`startPhoneVerification(phonenumber fields)` followed by `verify(phonenumber and user entered token)`

File: sms-authentication/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+## Setup the Authentication Function
+
+* Create a new Schema Extension Function and paste the schema from `schema-extension.graphql` and code from `sms-authentication.js`.
+* Create a new Permanent Access Token (PAT) in project settings. *It needs to have the same name as the function to make it available in the execution context of the function.*
+* Remove all Create permissions from the `User` type. The function uses a Permanent Access Token to create users via the API so the permissions are not needed.
+
+## Twilio App Setup
+
+Copy the `TWILIO_SID` and `TWILIO_AUTH_TOKEN` over to your new inline function.
```
Comment (kbrandwijk) [inline sms-authentication/README.md]: How do you get these?

Comment (peterpme) [inline sms-authentication/README.md]: Good catch, thanks!

**Comment by peterpme** [comment]
@kbrandwijk thanks for the review! I had no idea that twilio offers phone verification. That is awesome!!! I'll have to look into that :)

Everything is done on the server (schema extension). The client passes in:
- phone number (first screen)
- confirmation code (second screen)

File: sms-authentication/README.md [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+3. Your app calls the Graphcool mutation `authenticateSmsUser(phone: String!)`
+4. The `authenticateSmsUser` function checks if the user exists, if not, a new `User` will be created
+5. The app will generate an SMS code and a JWT token that will be saved under `User`
+6. Twilio sends that sms code to the phone number provided
+7. Your `User` model is updated with the JWT-encrypted sms code
+8. Your user will receive a text message. Your app will take that sms code, unencrypt it and check to make sure its correct
+
+## Setup the Authentication Function
+
+* Create a new Schema Extension Function and paste the schema from `schema-extension.graphql` and code from `sms-authentication.js`.
```
Comment (kbrandwijk) [inline sms-authentication/README.md]: I think setting up the confirm-code SE is missing here, which is why I didn't understand the flow right away...

File: sms-authentication/README.md (hunk 1 above)
Comment (kbrandwijk) [inline sms-authentication/README.md]: Maybe add a step to the Twilio App Setup to register the app on Twilio, and where to find the required keys.

File: sms-authentication/README.md (hunk 2 above)
Comment (peterpme) [inline sms-authentication/README.md]: gotcha, thank you!

File: sms-authentication/README.md (hunk 1 above)
Comment (peterpme) [inline sms-authentication/README.md]: Good call! thanks

**Comment by marktani** [comment]
Thanks for the PR @peterpme! Could you work in the changes you discussed with @kbrandwijk? Then I can merge it in 🙂

**Comment by peterpme** [comment]
Hey @kbrandwijk,

Let me know what you think!

I looked into Twilio phone number authentication and it uses a library called `authy` which is a 3rd party library that does much more than just sms authentication.

I'd be happy to add it but it seems like overkill for now.

There could be another guide if users are interested in one time passwords, one touch, phone intelligence, etc on top of phone verification?

If you're ok with the instructions then we should be able to merge this in :)

**Comment by peterpme** [comment]
@kbrandwijk mind giving me the 👍 so @marktani can merge this in?

Sideway: we discussed using Authy via Twilio to confirm sms token but given that authy.com has its own process @kbrandwijk will create another guide for Authy :smile:

Thanks!!!

**Comment by kbrandwijk** [comment]
@peterpme Looks fine to me. It's a very manual approach using the building blocks Twilio offers, but it works well. There's a FR in the issues list for a variation on this theme using the Authy components from Twilio, but it's out of scope for this example.

**Comment by marktani** [comment]
Did some styling and wording changes, @peterpme could you add a word about the `JWT_SECRET_KEY` constant? Is that obtainable from Twilio?

**Comment by marktani** [comment]
Thanks @peterpme! 🙌

**Pull request merged**